enzyme/Enzyme/ActivityAnalysis.cpp

```diff
@@ -318,114 +318,204 @@ const char *DemangledKnownInactiveFunctionsStartingWith[] = {
 
 };
 
-/// Is the use of value val as an argument of call CI known to be inactive
-/// This tool can only be used when in DOWN mode
-bool ActivityAnalyzer::isFunctionArgumentConstant(CallInst *CI, Value *val) {
-  assert(directions & DOWN);
-  if (CI->hasFnAttr("enzyme_inactive"))
-    return true;
-
-  Function *F = getFunctionFromCall(CI);
-
-  // Indirect function calls may actively use the argument
-  if (F == nullptr)
-    return false;
-
-  if (F->hasFnAttribute("enzyme_inactive")) {
+// Return true if the function given by called (possibly null), funcName
+// is  1) inactive (meaning no derivative info is propagated internally)
+// and 2) the return value is guaranteed to be inactive.
+//          This means the return cannot include mutable memory.
+bool isFunctionInactiveCommon(Function *called, StringRef funcName,
+                              llvm::TargetLibraryInfo &TLI) {
+  if (called && called->hasFnAttribute("enzyme_inactive")) {
     return true;
   }
-
-  auto Name = F->getName();
-
-  // Allocations, deallocations, and c++ guards don't impact the activity
-  // of arguments
-  if (isAllocationFunction(Name, TLI) || isDeallocationFunction(Name, TLI))
-    return true;
-  if (Name == "posix_memalign")
+  if (funcName == "free" || funcName == "_ZdlPv" || funcName == "_ZdlPvm" ||
+      funcName == "munmap") {
```

Comment:
You don't seem to return true for memalign anymore (and the other 4 are new). Is that intended?

Reply:
Per the comment the memalign allocates and thus isn't in the common code, but is in the inst variant.